### DISCUSSION TITLE
Validate Livewire properties before they are used in rendering

### DISCUSSION THREAD

**Discussion by martinszemitis** [livewire/livewire#2903] [1. Help] [answered]
I have datatable Livewire component with filters.

In render() method date fields from filters are used to query data.

When I enter wrong date format, render() method fails because following code fails:
```
Contact::query()
    ->when($this->filters['created_from'], fn($query, $date) => $query->where('created_at', '>=', Carbon::createFromFormat($this->dateFormat['php'],$this->filters['created_from'])))
```

I have created validation for these fields in **updated** hook, but seems that Livewire still executes render() if validations fail or it is even executed before updated hook.

It seems that validation doesn't work on **updating** Livewire hooks, probably because validation uses public properties directly and they are not updated yet.

What would be the best approach to avoid issue with wrong date formats failing rendering?

Is the only option to have one more data structure for validated filter fields or do you see some other solution? Or am I doing something wrong in general?

**Comment by drbornot** [comment]
Right now I'm building some filters datatable like you with multiples selects element filtering and star and end dates elements. One of the solution I made was related with your post. I did the following
```
public $validDates = true;
public $start_date, $end_date;

public function render()
{

    return view('livewire.blade-component', [
            'collections' => $this->validDates ? $this->filterBySelectedInputs() : Collection::empty()->paginate(10)
        ]);
}

public function updatedStartDate($value)
{
     $validator = Validator::make([
            'start_date' => $this->prepareDatesForValidation($value,'start'),
            'end_date' => $this->prepareDatesForValidation($this->end_date,'end')
        ],[
            'start_date' => 'date|before:end_date',
            'end_date' => 'date|after:start_date'
        ]);
        if ($validator->fails()) {
            $this->validDates = false;
            $validator->validate();
        }
        else {
            $this->validDates = true;
        }
}

public function updatedEndDate($value)
{
        $validator = Validator::make([
            'start_date' => $this->prepareDatesForValidation($this->start_date,'start'),
            'end_date' => $this->prepareDatesForValidation($value,'end')
        ],[
            'start_date' => 'date|before:end_date',
            'end_date' => 'date|after:start_date'
        ]);
        if ($validator->fails()) {
            $this->validDates = false;
            $validator->validate();
        }
        else {
            $this->validDates = true;
        }
}

public function prepareDatesForValidation($value,$arg)
{
      if ($arg == 'start')
         $this->start_date = date_create($value);
      elseif ($arg == 'end')
         $this->end_date = date_create($value . ' 23:59:59');	
}
```
This check if validation fails, if fails set validDates property to false and then render an empty collection and show the errors message on dates elements.

**Reply by martinszemitis** [reply]
@drbornot thanks for posting your solution! Yes that works, but that leads to one additional property for filter field or group of fields.

I actually found quite elegant solution - could be that you and someone else in the future could used it:
```
public function updating($field, $value)
    {

        if(Str::before($field, '.') === 'filters') {
            $validator = match ($field) {
                'filters.created_from' => Validator::make(
                    ['created_from' => $value],
                    ['created_from' => 'nullable|date_format:"d/m/Y"'],
                ),
                'filters.created_to' => Validator::make(
                    ['created_to' => $value],
                    ['created_to' => 'nullable|date_format:"d/m/Y"'],
                ),
            };
            $validator->validate();
        }
    }
```

I created custom validator in **updating** method. If validation fails, Livewire property is not updated and validation error is shown. This way I don't have to check anything during rendering - data are validated before rendering.

**Comment by martinszemitis** [comment, marked as answer]
*I will put my comment as an answer to be able to close this question and mark this as best solution (at least I think so :) )*

I actually found quite elegant solution - could be that you and someone else in the future could used it:

```
public function updating($field, $value)
    {

        if(Str::before($field, '.') === 'filters') {
            $validator = match ($field) {
                'filters.created_from' => Validator::make(
                    ['created_from' => $value],
                    ['created_from' => 'nullable|date_format:"d/m/Y"'],
                ),
                'filters.created_to' => Validator::make(
                    ['created_to' => $value],
                    ['created_to' => 'nullable|date_format:"d/m/Y"'],
                ),
            };
            $validator->validate();
        }
    }
```
I created custom validator in **updating** method. If validation fails, Livewire property is not updated and validation error is shown. This way I don't have to check anything during rendering - data are validated before rendering.
